interface/src/app/analyze/analysis/sampleBin.js

```diff
@@ -77,6 +80,26 @@ function($log, $cacheFactory, $q, Sample, Activity) {
       }
     },
 
+    getSampleGroups: function() {
+      var keys = Object.keys(this.sampleGroups);
+      var samplesByGroup = {};
+      var i;
+
+      // each distinct value in sampleGroups becomes a key in samplesByGroup,
+      // and the keys of sampleGroups are collected in a list within each
+      // corresponding value of samplesByGroup
+      for (i = 0; i < keys.length; i++) {
+        if (samplesByGroup[this.sampleGroups[+keys[i]]]) {
```

Comment:
I went with a slightly different approach that I think helps make this murky code block even more clear: the ambiguous `sampleGroups` is now called `sampleToGroup` to emphasize its function, and the method `getSampleGroups` is now called `getSamplesByGroup`. I also used a temp var in the loop to clarify what that was. I think it's much easier to follow now.